VnodeCacheDesign.md

```diff
@@ -0,0 +1,55 @@
+# Vnode Cache Design
+
+## Goals
+
+- Reduce the time required to determine if a file is inside of a virtualization root
+- Reduce the time required to find a file's virtualization root
+- **TBD:** Eliminate the use of file flags to determine if a file is inside of a root
+
+## Design 
+
+### Cached Data
+
+PrjFSKext will cache the following information for each vnode:
+
+- The `vnode_t` itself
+- The vnode's `uint32_t` vid (generation number)
+- `VirtualizationRootHandle` for the vnode (or `RootHandle_None` if the vnode is not in a root).  **NOTE** This assumes that `VirtualizationRootHandle`  are not recycled as VFS4G instances are unmounted\remounted.
+
+The data for each vnode will be stored in a hash table that assumes there will be `desiredvnodes` total vnodes.
+
+### Using the Cache
+
+When PrjFSKext needs to determine the virtualization root of a file (or if a file is inside of a virtualization root at all), it will look up the vnode in its cache.
+
+- If the vnode exists in the cache and its `vid` has not changed, PrjFSKext will use the `VirtualizationRootHandle` in the cache.
+
+- If the vnode exists in the cache and its `vid` has changed, PrjFSKext will lookup its `VirtualizationRootHandle` (using the existing code in PrjFSKext) and update the cache entry.
+
+- If the vnode does not exist in the cache,  PrjFSKext will lookup its `VirtualizationRootHandle` (using the existing code in PrjFSKext) and insert the vnode into the cache.
+
+### Special Case: Renames
+
+*Directory renames* 
+
+Directory renames are a special case because all of the directory's children may need to have their `VirtualizationRootHandle` updated. Rather than recording\tracking all of the parent\child relationships among the vnodes PrjFSKext will simply invalidate the entire cache when it detects that a directory has been renamed in\out of a virtualization root.  If PrjFSKext is unable to determine if the rename was within the same root (e.g. the directory's vnode was recycled) then PrjFSKext will invalidate the cache.
+
+*File renames*
+
+When a file is renamed PrjFSKext will update its `VirtualizationRootHandle` and\or `vid` as needed.
+
+## Future Enhancements
+
+- The vnode cache could track which vnodes have received the `KAUTH_FILEOP_WILL_RENAME` file operation (on Mojave) so that when handling `KAUTH_VNODE_DELETE` the kext can know if the delete vnode operation is for a rename.
+
+## Issues not addressed by the cache
+
+- Improving the speed and reliability of finding a vnode's path
+- Addressing failures to get a vnode's parent (when the vnode is not in the cache)
+- Finding the roots for hard links that span repos (and general handling of hard links that cross repo boundaries)
+
+## Open Questions
```

Comment:
My biggest open question is: what is the design of the hash table? As we've discussed, we don't need a full-fledged hash table since we want to store every vnode, but we do need to at least implement a hash function and index collisions. What does that look like?

Reply:
Related: what thread-safety guarantees do we need?

Reply:
Great questions:

> we do need to at least implement a hash function 

Based on what I've read for pointers it's common to simply use the value of the pointer, or the value right-shifted by 2 to 3 bits.  I can test out some different options and see how they perform. 

> and index collisions

Index collisions will be handled by adding a new entry to the list in each bucket.  We can compare `vnode_t` values to know when we've found a match (when walking the list).

> what thread-safety guarantees do we need

Another great question, I'll add more specifics in this area to the doc.

Reply:
> Index collisions will be handled by adding a new entry to the list in each bucket. 

That assumes that you're implementing buckets. Certainly a valid hash table design choice, but seems to be more complexity than we'd need here, and also potentially wasteful on memory.

We should at least consider other options (linear probing? secondary hash?). Perhaps we'll find that the pointer value hashes cluster more than we'd like and we do want the buckets; in that case, we may not want to preallocate the `desiredvnodes` entries up front, and then have the complexity of needing to reallocate in some scenarios.

All of that is to say - I think we need more design detail on the hash table implementation :-)

Reply:
>  Certainly a valid hash table design choice, but seems to be more complexity than we'd need here, and also potentially wasteful on memory.

Linear probing has some big advantages: no memory allocations and locality of reference in a collision scenario. In this application it's probably a great choice.

Reply:
I'd definitely avoid the linked list approach here, we want to avoid memory allocations while holding the hash table lock, which makes the insertion code very complex. Linked lists are also awful for locality of reference/cache behaviour. So I'd almost certainly go with open addressing if we're talking hash tables. Linear probing ought to be good enough if the hash function is decent.

Speaking of hash functions: don't forget they're not arbitrary pointers, as vnodes use the zone allocator: If I'm reading the output of `zprint` correctly, `vnode`s are 248 bytes on macOS 10.13.6, and allocated in 8KiB slabs. (33 vnodes per slab) So each vnode pointer will have one of only 33 values in the low 13 bits (uniformly distributed), a whole bunch more non-uniform entropy in bits 13 and up, at least to a point; there will probably be very little entropy in the upper ~32 bits. Anyway, this sort of pattern is poison for the identity hash function if using power-of-2 capacities, but it really shouldn't be hard to come up with a hash function that works well. A good start would be to collect all vnode pointer values we see during a series of intensive test runs (with reboots) and dump them to a file we can then use as input for offline tests of hash functions & tables.

This is assuming we go with a hash table. Given the bit pattern, we could also go with a radix/prefix tree. Those are especially good if memory use is dominated by the key; in our case, each table entry is fairly large compared to the key, so the memory saving from not storing the vnode pointer itself in each table entry is negligible. On the other hand, we don't have to contend with load factors, collision resolution, etc.

Given how critical this data structure is to performance, it might be worth trying a few different options.

Reply:
> If the position in the prefix tree already encodes the vnode pointer, we can more or less halve the memory cost.

What pieces of data were you thinking would be recorded to encode portions of the pointer address while moving down the tree?  In other words, how would each node record what chunks of the pointer's address are encoded by each of its children (or how would the children themselves record this information)?

The options that I've come up with have all been pretty inefficient.

Reply:
This is a very quick & dirty sketch of the sort of thing I was thinking of: https://gist.github.com/pmj/a25abf621b830ba9d7f7dee6d6959d15

There are no doubt some optimisations to be made, and the insertion logic is, uh, *left as an exercise to the reader*, but I **think** it should be possible to turn it into something practical. Whether it's faster than a hash table remains to be seen.

Reply:
> This is a very quick & dirty sketch of the sort of thing I was thinking of: 

Thanks!

> Whether the approach is faster than a hash table remains to be seen.

Agreed, it's looking like the big takeaway is we'll need to implement and perf test a few options to see what's best.